plugins/glustershd/transactions.go

```diff
@@ -0,0 +1,112 @@
+package glustershd
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/gluster/glusterd2/glusterd2/brick"
+	"github.com/gluster/glusterd2/glusterd2/daemon"
+	"github.com/gluster/glusterd2/glusterd2/servers/sunrpc/dict"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+
+	log "github.com/sirupsen/logrus"
+)
+
+func getHxlChildrenCount(volinfo *volume.Volinfo) (int, string) {
+	if volinfo.Type == volume.Replicate || volinfo.Type == volume.DistReplicate {
+		return volinfo.Subvols[0].ReplicaCount, "replicate"
+	}
+	return volinfo.Subvols[0].DisperseCount, "disperse"
+}
+
+func addHxlatorToDict(reqDict map[string]string, volinfo *volume.Volinfo, index int, count int, xlType string) map[string]string {
+	key := fmt.Sprintf("xl-%d", count)
+	xName := fmt.Sprintf("%s-%s-%d", volinfo.Name, xlType, index)
+	reqDict[key] = xName
+	reqDict[xName] = strconv.Itoa(index)
+	return reqDict
+}
+
+func selectHxlatorsWithBricks(volinfo *volume.Volinfo, healType string) map[string]string {
+	index := 1
+	hxlatorCount := 0
+	reqDict := make(map[string]string)
+	reqDict["heal-op"] = healType
+	reqDict["xl-op"] = reqDict["heal-op"]
+	reqDict["volname"] = volinfo.Name
+	reqDict["sync-mgmt-operation"] = "20"
+	reqDict["vol-id"] = volinfo.ID.String()
+	hxlChildren, xlType := getHxlChildrenCount(volinfo)
+	volBricks := volinfo.GetBricks()
+	for brick := range volBricks {
+		hostKey := fmt.Sprintf("%d-hostname", index-1)
+		brickPathKey := fmt.Sprintf("%d-path", index-1)
+		reqDict[hostKey] = volBricks[brick].Hostname
+		reqDict[brickPathKey] = volBricks[brick].Path
+
+		if index%hxlChildren == 0 {
+			reqDict = addHxlatorToDict(reqDict, volinfo, (index-1)/hxlChildren, hxlatorCount, xlType)
+			hxlatorCount++
+		}
+		index++
+	}
+	reqDict["count"] = strconv.Itoa(hxlatorCount)
+	return reqDict
+}
+
+func txnSelfHeal(c transaction.TxnCtx) error {
+
+	var volinfo volume.Volinfo
+	if err := c.Get("volinfo", &volinfo); err != nil {
+		c.Logger().WithError(err).WithField(
```

Comment:
Remove these log entries. The txn context code already does this. See `Get()` in `glusterd2/transaction/context.go`